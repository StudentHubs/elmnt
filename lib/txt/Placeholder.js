"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var refluent_1 = require("refluent");
var utils_1 = require("../utils");
exports.default = refluent_1.default
    .yield(function (_a) {
    var text = _a.text, value = _a.value, prompt = _a.prompt, next = _a.next;
    return prompt
        ? next(function (p) { return (__assign({}, p, { text: text
                .split('\n')
                .map(function (l, i) { return (value.split('\n')[i] ? '' : l); })
                .join('\n') })); })
        : text && !value
            ? next()
            : null;
})
    .do(utils_1.restyle(function (style) {
    return style.merge({
        position: 'absolute',
        top: 0,
        left: 0,
        width: '100%',
        height: '100%',
        padding: 0,
        display: 'block',
        whiteSpace: 'pre-wrap',
    });
}))
    .yield(function (_a) {
    var text = _a.text, style = _a.style;
    return React.createElement("span", { style: style }, text);
});
