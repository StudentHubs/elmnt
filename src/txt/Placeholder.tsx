import * as React from 'react';
import m, { Comp, restyle } from 'mishmash';

export interface PlaceholderProps {
  text?: string;
  value: string;
  prompt?: boolean;
  style: React.CSSProperties;
}
export default m
  .branch(
    ({ prompt }) => prompt,
    m.map(({ text, ...props }) => ({
      ...props,
      text: text
        .split('\n')
        .map((l, i) => (props.value.split('\n')[i] ? '' : l))
        .join('\n'),
    })),
    m.branch(({ text, value }) => !text || value, m.render()),
  )
  .map(
    restyle([
      [
        'merge',
        {
          position: 'absolute',
          top: 0,
          left: 0,
          width: '100%',
          height: '100%',
          padding: 0,
          display: 'block',
          whiteSpace: 'pre-wrap',
        },
      ],
    ]),
  )(({ text, style }) => <span style={style}>{text}</span>) as Comp<
  PlaceholderProps
>;
