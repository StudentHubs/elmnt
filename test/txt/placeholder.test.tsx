import * as React from 'react';
import { render } from 'enzyme';

import Placeholder from '../../src/txt/Placeholder';

describe('txt: placeholder', () => {
  it('no text, empty value', () => {
    expect(
      render(<Placeholder value="" style={{ background: 'red' }} />),
    ).toMatchSnapshot();
  });

  it('text, empty value', () => {
    expect(
      render(
        <Placeholder text="Hello!" value="" style={{ background: 'red' }} />,
      ),
    ).toMatchSnapshot();
  });

  it('text, value', () => {
    expect(
      render(
        <Placeholder
          text="Hello!"
          value="World!"
          style={{ background: 'red' }}
        />,
      ),
    ).toMatchSnapshot();
  });
});
