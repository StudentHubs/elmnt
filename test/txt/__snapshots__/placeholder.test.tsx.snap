// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`txt: placeholder no text, empty value 1`] = `null`;

exports[`txt: placeholder text, empty value 1`] = `
<span
  style="background:red;position:absolute;top:0;left:0;width:100%;height:100%;padding:0;display:block"
>
  Hello!
</span>
`;

exports[`txt: placeholder text, value 1`] = `null`;
